### PR TITLE
Telesci

### PR DESCRIPTION
**Pull request by BlueNothing**
Added: Allows for the construction of rapid crate senders, telepads, and telescience computers in RnD and reimplements

[why]: Telescience code restoration doesn't help without devices that execute it, and this allows those devices to be made in a way that reduces strain on mappers by allowing them to be made on an ad-hoc basis by science staff.

### CODE REVIEW

**Comment by BlueNothing** [comment]
What does it mean by saying that all checks have failed, Travis CI, etc? Sorry, I'm pretty new at this.

**Comment by SgtxRootbeer** [comment]
It means your build failed to compile

**Comment by BlueNothing** [comment]
Oh. Damn. Any idea why?

**Comment by SgtxRootbeer** [comment]
Ya, check the details on Travis, but in short, check the image I sent
screenshot_20170903-154032

**Comment by BlueNothing** [comment]
It can't *open* those two .dm files? Huh.

**Comment by p440** [comment]
its because the files dont exist. you must have forgot to commit them or something, i couldnt even find the modules/telesci folder

**Comment by BlueNothing** [comment]
I'll get around to fixing that when I get back.

**Comment by BlueNothing** [comment]
How do I add files to a commit?

**Comment by BlueNothing** [comment]
Added the requisite module in another PR, so I'll need someone to merge #30388 and #30393, but it should compile then.

**Comment by SgtxRootbeer** [comment]
This seems like the wrong way to be going about this fix, however, I love telescience, so I approve

**Comment by BlueNothing** [comment]
I'm a total noob at this, or I'd do it the right way.

**Comment by BlueNothing** [comment]
Wait... how *do* I do it the right way?

**Comment by BlueNothing** [comment]
By 'wrong way' are you referring to my kludging it together in two PRs, or the actual code in said PRs?

**Comment by SgtxRootbeer** [comment]
Add the files to the "BlueNothing:Telesci" branch of your repository

**Comment by SgtxRootbeer** [comment]
And yes, you shouldn't use 2 PRs if they BOTH require code from each other to work.
(Because then they cannot be merged separately, unlike an author splitting PRs because they are unrelated and the content of one doesn't depend on the other ex: a bugfix and a feature being separated into 2 PRs.)

**Comment by BlueNothing** [comment]
Okay, now it should work... pending Travis.

**Comment by BlueNothing** [comment]
Yes!

**Comment by BlueNothing** [comment]
@SgtxRootbeer Thanks! It passes Travis now.

**Comment by vcordie** [comment]
As the person who originally removed telesci, it was removed for being overpowered as fuck with no real way to fix it short of completely preventing it from working on the station z level

**Comment by BlueNothing** [comment]
What were the benefits of it? What purpose did it serve?

**Comment by BlueNothing** [comment]
Just trying to figure out what motivation you understood it to have for existing, whether it filled that niche, and whether that niche has yet to be filled effectively by something else.

**Comment by vcordie** [comment]
Okay let me put it to you this way
Whenever I was rd, if the round was nukeops and they didn't bomb sci, they'd lose
If I was traitor, within 13 minutes sec, engineering, would cease to exist without me ever leaving rnd
if I had a theft objective, within 11 minutes it'd be completed with no way to stop me.
I could steal the armory without ever entering security

The niche, good purpose it served was teleporting dead miners off lavaland directly into cloning.
Too bad it came with a whole bunch of other shit.

**Comment by BlueNothing** [comment]
You must've been pretty good at doing it fast.

**Comment by BlueNothing** [comment]
Though, nowadays, you can still set up bombs on mining with quantum pads, the nuke-ops thing could be fixed by making it so that bombs can't teleport intra-z-level unless the machine is emagged, or by making it so that the disk can't be teleported. Unfortunately, I don't know how to do that, but this is a start. The armory thing is a side effect of working-as-intended, and is likely to remain so. Likewise with most theft.

**Comment by XDTM** [comment]
My issue with it was that it was overcomplicated and favored people who used an external tool.

I tried filling the niche a bit with launchpads, but they came out too weak to be useful. 
Maybe buffing them might be better than reintroducing all the issues with telesci.

**Comment by FantasticFwoosh** [comment]
Keep the rapid crate senders, i want to see if it will make cargo marginally more effective as well as self organise ore box transporting stations between lavaland & the station.

I forget how they used to work.

**Comment by SgtxRootbeer** [comment]
You should take vcordie's statement with a grain of salt since they got it removed by breaking the rules, which already prohibits metagaming the nukeops shuttle location.
The role it filled IMO was 
1, a thematic use of bluespace tech. Considering how much we know about bluespace, being able to teleport people anywhere seems to be something we should be able to do. (Also, it's like a transporter from star trek)
2, it was a needed tool for destroying bases of cults, ai, etc. In it's absence, making bases for any cult is now an ez pz way to amplify your cults remaining manpower without much of a trade off.

To fix the in-game vs out-of-game tool problem, we can make it work off of offsets or putting in a destination coordinate, or, we can give cargo/rnd a new item, which is the calculator except it's in-game. Call it a bluespace telescope or something.

**Comment by Grazyn** [comment]
@vcordie 
let's be honest here, you abused it pretty hard especially with nuke ops, sending gps to every preset op shuttle location to find it and then bomb it, you got banned for it and removed telesci out of spite. People had been using telesci for years, doing that kind of stuff from time to time (remote telebombing, stealing pins from armory etc.) then you came and started doing it literally every single fucking round running the gimmick into the ground until people were happy to see it gone.

I'm not saying it was a good feature, it had the potential to be abused and it required an outside tool which is already a bad thing, but it wasn't really a problem until you came along so trying to play the hero we needed who saved us from a broken overpowered feature really rustles my jimmies

**Comment by BlueNothing** [comment]
I'd say either a coordinate system or in-gaming the calculator would be
brilliant solutions, especially since telescience is a late-game capstone
kind of thing like jaws of life or xray lasers.

On Sep 4, 2017 9:39 AM, "Grazyn" <notifications@github.com> wrote:

@vcordie <https://github.com/vcordie>
let's be honest here, you abused it pretty hard especially with nuke ops,
sending gps to every preset op shuttle location to find it and then bomb
it, you got banned for it and removed telesci out of spite. People had been
using telesci for years, doing that kind of stuff from time to time (remote
telebombing, stealing pins from armory etc.) then you came and started
doing it literally every single fucking round running the gimmick into the
ground until people were happy to see it gone.

I'm not saying it was a good feature, it had the potential to be abused and
it required an outside tool which is already a bad thing, but it wasn't
really a problem until you came along so trying to play the hero we needed
who saved us from a broken overpowered feature really rustles my jimmies

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/30388#issuecomment-326965971>,
or mute the thread

**Comment by Grazyn** [comment]
We already have a coordinate system, people using telesci usually relied on external maps with the coords for every tile (or just opened the z-level map file in dream maker). Adding the same system in game (with a "map console" for example) won't actually fix the problem, if anything it would make it easier to abuse.

**Comment by BlueNothing** [comment]
Can confirm, I did, on occasion, use the z-level mapfile solution back when. Still, as it stands, the external system just makes it friendly to the people who used it before on a level that makes them (since they've usually had some opportunities to improve their skills since then) even more alarming. By moving it to an internal system, or using a coordinate-plus-linear-x-and-y-offset system, it levels the playing field somewhat. So abuse is to be expected as an extension of use, but I suspect that's the case with mostly anything. Pros will still do well with an easier system, but newer users will have an easier time with the revisions suggested, which will balance out the user list somewhat, at least.

**Comment by BlueNothing** [comment]
@XDTM That's a good idea, and in the end, maybe that'll be what happens, or we'll wind up somewhere between the two cases. At the worst, I've at least sparked a discussion on the matter.

**Comment by BlueNothing** [comment]
@FantasticFwoosh The likely behavior of Rapid Crate Senders will probably overlap somewhat with the mining extraction kits, except for the fact that they require RnD resource investments, and work more generally on account of not requiring the 'must be outside' condition to hold. They cost bluespace, though, so they require a significant supply of diamonds or bluespace crystals. It won't exactly be surprising if the tide uses them as a ghetto hand-tele with lockers.

**Comment by vcordie** [comment]
I actually was going to pr a removal of telesci prior to being banned, but
was too busy playing to do it. I was actually talking about doing it for a
week prior.
The fact that I removed it after being banned does not change the fact that
it's a shit feature that only existed to be abused through the use of
external tools

**Comment by BlueNothing** [comment]
Then, in time, let's at least make it so that if it exists to be abused, the tools are internal... but first, let's see if the playerbase wants it at all. So far, from what little sampling I did, feedback on the idea seemed favorable.

**Comment by sawrge** [comment]
if you asked the playerbase if they wanted to all spawn with pulse cannons they'd say yes. gauging whether or not you should re-add garbage that was removed for a reason based on vocal minorities clamoring for it with no regard for the game's health is deadass tarded

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Supermichael777** [comment]
the problem is that its limitations were soft, it needed math. however excel exists

add a super obvious overlay for the target and a delay, (30) seconds should do. Optionally i would think messages about what objects are about to arrive and penalties for trying to use it too often should be good. And teleportable object blacklists if it starts getting out of hand again. vcordie was faced with his toy getting nerfed so he smashed it.

**Comment by BlueNothing** [comment]
I agree that blacklists would be a good extension. Overlays on target seem reasonable, and some kind of info (maybe an overlay picture of what's being teleported in, or a common overlay for trying to pull something in) would be a good elaboration on that. I don't think a delay is especially helpful, since it removes the possibility of trying to teleport a mob target even beyond the difficulty of getting a moving target to be in the right place.

File: code/game/machinery/quantum_pad.dm [hunk 1]
```diff
@@ -15,6 +15,20 @@
 	var/teleporting = 0 //if it's in the process of teleporting
 	var/power_efficiency = 1
 	var/obj/machinery/quantumpad/linked_pad
+	
+	//mapping
+	var/static/list/mapped_quantum_pads = list()
+	var/map_pad_id = "" as text //what's my name
```
[the remark this answers is not in the document]
Comment (BlueNothing) [reply code/game/machinery/quantum_pad.dm]: At the end of the list of vars, or...

**Comment by SgtxRootbeer** [comment]
we should have all teleporting be instant, but if you wish to make blacklists, or something of the like, limit telepad operation to mecha and mobs only. allow stuff that the mob/mecha is carrying to be teleported too (obviously), this creates a small dependency on robotics to build a ripley because its clamps can load unachored stuff that is too large to be carried into its cargo compartment (SM shard, crates, unachored machinery, etc.)

**Comment by BlueNothing** [comment]
I don't think it needs to go *that* far, necessarily, just a small *blacklist* (not a whitelist), on things like, say, the nuke, the disk... any other stuff that's similarly stupidly unwise to allow to teleport?

**Comment by SgtxRootbeer** [comment]
Yes, but a magical blacklist of items that cant be teleported is lame. (speaking of, dominators and arks are anchored objects, and as a result, cannot be teleported regardless)
if we limit to mobs and mecha, it allows people to teleport any item, but it puts the teleporting team at risk, and limits how many items they can teleport at a time, and encourages co-worker cooperation.

**Comment by BlueNothing** [comment]
I don't think the list would be very long to begin with. Aside from those
five items, I couldn't think of much else.

File: code/game/machinery/quantum_pad.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (BlueNothing) [reply code/game/machinery/quantum_pad.dm]: It looks like those might've been there from before?

[the remark this answers is not in the document]
Comment (Cyberboss) [reply code/game/machinery/quantum_pad.dm]: At the end of the list

**Comment by BlueNothing** [comment]
Did that do it?

**Comment by BlueNothing** [comment]
I fixed the formatting, per your request, but this action created a merge conflict. Evidently, the poor formatting is present in the code as it exists.

**Comment by BlueNothing** [comment]
@Cyberboss The formatting error you object to is actually a carry-over from the current file: https://github.com/tgstation/tgstation/blob/master/code/game/machinery/quantum_pad.dm

**Comment by BlueNothing** [comment]
@Cyberboss  Merge conflict resolved, hopefully to your liking.

**Comment by optimumtact** [comment]
I don't think any maintainer will merge this without substantial changes to telesci to make it far less powerful, my original suggestion was making it only able to teleport people around, and not items.

**Comment by lzimann** [comment]
Let's not make the mistake of re-adding telesci the way it was before. If you really want it you will have to come with something new. PR denied, sorry.

**Comment by BlueNothing** [comment]
Why would old-telesci be a mistake? Also, I thought of a potential balancing item. Some kind of telescience blackout beacon, probably start out with two in the armory, so you couldn't telesci out the armory without moving the beacons into another room first, and set it up so cargo can order more. Basically, the way those would work is that the telesci code would check whether the room targeted had a beacon in it of that type, and if so, it wouldn't teleport. How reasonable would something like that be?

Regarding the state of telescience: Quantum pads are useful, but rather limited, and require significant setup, so most rounds will see exactly one pair set up, maybe a total of three pads on a loop (RnD->Mining->Emergency Shuttle->RnD). Bluespace launchpads are even worse, with basically nobody using them (even the more versatile *antag* version sees fairly little use from my anecdotal experience). So while maybe reinserting telescience from whole cloth is a bit excessive, what would your thoughts be on the Rapid Cargo Sender, and maybe augmenting the launchpad somehow so that it actually sees use? That, or maybe just removing it, because it seems like most teleportation shenanigans are either going to be so situational as to be useless or never used, or so anti-situational as to be overpowered.

**Pull request closed without merging**